### ISSUE TITLE
[rllib]: Which losses exactly are used in RLlib's PPO algorithm?

### ISSUE THREAD

**Issue by adriendoerig** [ray-project/ray#13837]
I am struggling to understand how the PPO algorithm in RLlib works. Specifically, the RLlib documentation (https://github.com/ray-project/ray/tree/master/rllib/agents/ppo) says that there are two versions: one based on the TRPO idea and one based on the standard PPO clipped surrogate objective (which I want to use). However, I cannot find where or how to pick which implementation to use.

In addition, there seems to be something going on with KL divergence, specifically using KLCoeffMixin (defined here: https://github.com/ray-project/ray/blob/master/rllib/agents/ppo/ppo_tf_policy.py). It seems to be updating the weight of a KL term in the loss. This may make sense for the TRPO version, but not for the surrogate objective version.

Can you help me understand how exactly the PPO algorithm works in RLlib?

**Comment by ericl** [comment]
Both are used (both clipping and KL constraint).

**Comment by adriendoerig** [comment]
Hi, thanks for the reply! 

I know "both are used", that's what the documentation says. But you mean they are used simultaneously? This is not standard practice, from my understanding. Usually, _either_ we use the TRPO version _with_ the KL divergence term, _or_ we use the clipped gradient version _without_ the KL divergence term.

I guess my problem is that I don't understand what the PPO loss function is in RLlib. Do you think you could right it down? This would greatly clarify. Thanks!

**Comment by sven1977** [comment]
Yes, they are both used. The weight of the KL-loss term is determined by the `kl_coeff` config parameter (which will be decreased/increased during training to reach the `kl_target` value). You can set `kl_coeff` to 0.0 if you don't want to use any KL-loss term.
